lib/awful/placement.lua

```diff
@@ -29,6 +29,132 @@ end
+-- methods.
+-- @param obj An object
+-- @tparam table args the method arguments
+-- @tparam[opt=nil] table new_geo A new geometry to replace the existing one
+-- @tparam[opt=false] boolean ignore_border_width Ignore the border
+-- @treturn table A table with *x*, *y*, *width* and *height*.
+local function geometry_common(obj, args, new_geo, ignore_border_width)
+    -- It's a mouse
+    if obj.coords then
+        local coords = new_geo and obj.coords(new_geo) or obj.coords()
+        return {x=coords.x, y=coords.y, width=0, height=0}
+    elseif obj.geometry then
+        local geo = obj.geometry
+
+        -- It is either a drawable or something that implement its API
+        if type(geo) == "function" then
+            local dgeo = area_common(obj, new_geo, ignore_border_width)
+
+            -- Apply the margins
+            if args.margins then
+                local delta = type(args.margins) == "table" and args.margins or {
+                    left = args.margins , right  = args.margins,
+                    top  = args.margins , bottom = args.margins
+                }
+
+                return {
+                    x      = dgeo.x      + (delta.left or 0),
+                    y      = dgeo.y      + (delta.top  or 0),
+                    width  = dgeo.width  - (delta.left or 0) - (delta.right  or 0),
+                    height = dgeo.height - (delta.top  or 0) - (delta.bottom or 0),
+                }
+            end
+
+            return dgeo
+        end
+
+        -- It is a screen, it doesn't support setting new sizes.
+        return a_screen.get_bounding_geometry(obj, args)
+    else
+        assert(false, "Invalid object")
+    end
+end
+
+--- Get the parent geometry from the standardized arguments API shared by all
+-- `awful.placement` methods.
+-- @param obj A screen or a drawable
+-- @tparam table args the method arguments
+-- @treturn table A table with *x*, *y*, *width* and *height*.
+local function get_parent_geometry(obj, args)
+    if args.bounding_rect then
+        return args.bounding_rect
+    elseif args.parent then
+        return geometry_common(args.parent, args)
+    elseif obj.screen then
+        return geometry_common(obj.screen, args)
+    else
+        return geometry_common(capi.screen[capi.mouse.screen], args)
+    end
+end
+
+--- Convert a rectangle and matrix coordinates info into a point.
+-- This is useful along with matrixes like `corners3x3` to convert
+-- indices into a geometry point.
+-- @tparam table geo a geometry table
+-- @tparam number corner_i The horizontal matrix index
+-- @tparam number corner_j The vertical matrix index
+-- @tparam number n The (square) matrix dimension
+-- @treturn table A table with *x* and *y* keys
+local function rect_to_point(geo, corner_i, corner_j, n)
+    return {
+        x = geo.x + corner_i * math.floor(geo.width  / (n-1)),
+        y = geo.y + corner_j * math.floor(geo.height / (n-1)),
```

Comment:
`2 * math.floor(1 / 2) = 0`, but `math.floor(2 * (1 / 2)) = 1`

Or, more generally, let width be an odd number: `2*math.floor(width / (3-1)) = 2*(width/2-0.5) = width-1 ~= width`